epitator/infection_annotator.py

```diff
@@ -0,0 +1,314 @@
+    # participle of "strike".
+    "VERB": {
+        "infection": ["infect", "sicken", "stricken", "strike", "diagnose", "afflict"],
+        "death": ["die"],
+        "hospitalization": ["hospitalize", "admit"]
+    }
+}
+
+
+def spacy_tokens_for_span(span):
+    doc = span.doc
+    if "spacy.tokens" not in doc.tiers:
+        doc.add_tier(SpacyAnnotator())
+    tokens_tier = span.doc.tiers["spacy.tokens"]
+    tokens = [t.token for t in tokens_tier.spans_contained_by_span(span)]
+    return(tokens)
+
+
+# FIXME: This should work better. Right now it removes possessives by checking
+# that a token.pos_ == "NOUN" is not token.dep_ == "poss". Previously, it was
+# nice and parsimonious, and just extracted the inner lemma dicts by pos_.
+# It'd be nice to have a consistently applicable approach, where, say, you
+# state a set of conditions and then state a set of lemmas to look for.
+# Another approach would be to break this out of the function and just apply
+# the appropriate lemmas at different points based on the code. Or to change
+# the point of abstraction such that you pass in a sequence of tokens and the
+# inner lemma dict, and then just match without checking for part of speech.
+def generate_attributes(tokens, attribute_lemmas=attribute_lemmas):
+    metadata = {}
+    attributes = []
+    for i, t in enumerate(tokens):
+        if t.pos_ == "NOUN" and t.dep_ == "poss":
+                continue
+        for category, lemmas in attribute_lemmas.get(t.pos_, {}).items():
+            if t.lemma_ in lemmas:
+                attributes.append(category)
+
+    metadata['attributes'] = attributes
+    return(metadata)
+
+
+def generate_counts(tokens, strict_only=False):
+    if len(tokens) == 0:
+        return {}
+    metadata = {}
+    metadata["attributes"] = []
+    quant_idx = [i for (i, t) in enumerate(tokens) if t.ent_type_ in ['CARDINAL', 'QUANTITY'] and t.dep_ == 'nummod']
+    if len(quant_idx) == 1:
+        count_text = tokens[quant_idx[0]].text
+        metadata["count"] = parse_count_text(count_text)
+
+    elif len(quant_idx) > 1:
+        # If we find multiple tokens, we deal with them thus: First, we group
+        # group consecutive tokens together in sub-lists: [1, 4, 5, 6, 8, 9]
+        # becomes [[1], [4, 5, 6], [8, 9]]. If there is only one of these
+        # groups, we process it as a single number. If there are multiple
+        # groups, we add them all to metadata["count"] as a list. This will
+        # cause InfectionAnnotator() to throw an error, however.
+        groups = []
+        for k, g in groupby(enumerate(quant_idx), lambda ix: ix[0] - ix[1]):
+            groups.append(list(map(itemgetter(1), list(g))))
+        if len(groups) == 1:
+            count_text = " ".join([tokens[i].text for i in groups[0]])
+            metadata["count"] = parse_count_text(count_text)
+            metadata["attributes"] = ["JOINED_CONSECUTIVE_TOKENS"]
```

Comment:
I would recommend creating a separate property for things that relate to the way the tokens were processed rather than the meaning of the tokens. Maybe you could call it "token_attributes."

Reply:
Agreed.